x-pack/plugins/stack_alerts/server/alert_types/geo_containment/es_query_builder.ts

```diff
@@ -0,0 +1,202 @@
+  }
+  return esFormattedQuery;
+};
+
+export async function getShapesFilters(
+  boundaryIndexTitle: string,
+  boundaryGeoField: string,
+  geoField: string,
+  callCluster: ILegacyScopedClusterClient['callAsCurrentUser'],
+  log: Logger,
+  alertId: string,
+  boundaryNameField?: string,
+  boundaryIndexQuery?: Query
+) {
+  const filters: Record<string, unknown> = {};
+  const shapesIdsNamesMap: Record<string, unknown> = {};
+  // Get all shapes in index
+  const boundaryData: SearchResponse<Record<string, unknown>> = await callCluster('search', {
+    index: boundaryIndexTitle,
+    body: {
+      size: MAX_SHAPES_QUERY_SIZE,
+      ...(boundaryIndexQuery ? { query: getEsFormattedQuery(boundaryIndexQuery) } : {}),
+    },
+  });
+
+  boundaryData.hits.hits.forEach(({ _index, _id }) => {
+    filters[_id] = {
+      geo_shape: {
+        [geoField]: {
+          indexed_shape: {
+            index: _index,
+            id: _id,
+            path: boundaryGeoField,
+          },
+        },
+      },
+    };
+  });
+  if (boundaryNameField) {
+    boundaryData.hits.hits.forEach(
+      ({ _source, _id }: { _source: Record<string, unknown>; _id: string }) => {
+        shapesIdsNamesMap[_id] = _source[boundaryNameField];
+      }
+    );
+  }
+  return {
+    shapesFilters: filters,
+    shapesIdsNamesMap,
+  };
+}
+
+export async function executeEsQueryFactory(
+  {
+    entity,
+    index,
+    dateField,
+    boundaryGeoField,
+    geoField,
+    boundaryIndexTitle,
+    indexQuery,
+  }: {
+    entity: string;
+    index: string;
+    dateField: string;
+    boundaryGeoField: string;
+    geoField: string;
+    boundaryIndexTitle: string;
+    boundaryNameField?: string;
+    indexQuery?: Query;
+  },
+  { callCluster }: { callCluster: ILegacyScopedClusterClient['callAsCurrentUser'] },
+  log: Logger,
+  shapesFilters: Record<string, unknown>
+) {
+  return async (
+    gteDateTime: Date | null,
+    ltDateTime: Date | null
+  ): Promise<SearchResponse<unknown> | undefined> => {
+    let esFormattedQuery;
+    if (indexQuery) {
+      const gteEpochDateTime = gteDateTime ? new Date(gteDateTime).getTime() : null;
+      const ltEpochDateTime = ltDateTime ? new Date(ltDateTime).getTime() : null;
+      const dateRangeUpdatedQuery =
+        indexQuery.language === 'kuery'
+          ? `(${dateField} >= "${gteEpochDateTime}" and ${dateField} < "${ltEpochDateTime}") and (${indexQuery.query})`
+          : `(${dateField}:[${gteDateTime} TO ${ltDateTime}]) AND (${indexQuery.query})`;
+      esFormattedQuery = getEsFormattedQuery({
+        query: dateRangeUpdatedQuery,
+        language: indexQuery.language,
+      });
+    }
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    const esQuery: Record<string, any> = {
+      index,
+      body: {
+        size: MAX_TOP_LEVEL_QUERY_SIZE,
+        aggs: {
+          shapes: {
+            filters: {
+              other_bucket_key: OTHER_CATEGORY,
```

Comment:
we _could_ remove having this other-bucket altogether, because we are explicitly filtering out these results later on:) . This may be hypothetically useful for more advanced fucntionalities later down the line (e.g. tracking vehicle-location through state)

idk, up2you if you would like to remove. It does negatively impact the query, creating buckets when they are not needed. in terms of code, footprint is minor.

Reply:
Yep, I considered this too. I only left it in because it was going to be the only query mod required for this PR and I happen to know I want this information for the forthcoming PR that adds better handling for if/when an alert instance goes into a resolved state. Basically it gives a firm "We've received an update and confirmed this is not in an area of interest" message. Agreed this is a performance hit in the meantime and should be removed if not used but I'm fairly confident it will be. Good to document this decision here though :+1: